File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
#         char_set = set()
#         l = 0
#         max_length= 0

#         for r in range(len(s)):
#             while s[r] in char_set:
#                 char_set.remove(s[l])
#                 l += 1
#             char_set.add(s[r])
#             max_length = max(max_length, r - l + 1)
#         return max_length
        if len(s)<=1:
            return len(s)
        char_set=set()
        left,right=0,0
        max_length=1
        while right<len(s):
            while s[right] in char_set:
                char_set.remove(s[left])
                left+=1

            char_set.add(s[right])
            max_length=max(max_length,right-left+1)
            right+=1
            
        return max_length
        
            

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
//     start left right point at index 0
//     define a set to track the char already included
//     with new char added to set, at the same time increase right point
//     with same char already in set, remove that char from set, and increase left pointer
    if (s.length<=1) return s.length
    let left=0
    let right=0
    let charSet=new Set()
    let maxLength=0
    
    while(right<s.length){
        
        while(charSet.has(s[right])){
            charSet.delete(s[left])
            left++
        }
        charSet.add(s[right]) 

        maxLength=Math.max(maxLength,right-left+1)
        right++
    }
    console.log(charSet)
    return maxLength
    
    
};